code/scripting/api/objs/ship.cpp

```diff
@@ -2104,6 +2104,17 @@ ADE_FUNC(warpIn, l_Ship, NULL, "Warps ship in", "boolean", "True if successful,
 	if(!objh->isValid())
 		return ADE_RETURN_NIL;
 
+	if (object_is_docked(objh->objp()))
+	{
+		// Ships that are docked need a designated dock leader to bring the entire group in.  The dock leader is, de facto, the arriving ship;
+		// so by scripting a certain ship to warp in, the script author has designated it as the leader.  That being said, if the script
+		// calls warpIn() multiple times on the same docked group, only set the flag on the first ship.
+		dock_function_info dfi;
+		dock_evaluate_all_docked_objects(objh->objp(), &dfi, dock_find_dock_leader_helper);
```

Comment:
This still leaves a corner case where `warpIn()` will mysteriously fail to do anything. Suppose we're in a pirate campaign, and in the first phase of the mission, the player is tasked with disabling a freighter and escorting a capture transport; in the second phase, the capture transport does a series of in-mission jumps to avoid pursuit, like BtA's "Shell Game". That will fail for no obvious reason if the freighter is initially carrying cargo. (I know that's pretty niche, but I _think_ it came up in an early draft of a BtA mission. I say "I think" because I was already souring on the design of that mission, so I didn't dig into it too closely; I'm going off a vague memory here.)

Removing the `Dock_leader` flag when a ship finishes arriving would avoid that issue (as well as a preexisting issue where two ships that are both dock leaders become (directly or indirectly) docked before `warpIn()` is called).

If that's not practical, we could just make the ship `warpIn()` is called on the dock leader (removing the flag from anything it's docked to), and document that `warpIn()` shouldn't be called on multiple ships in the same dock group.

Reply:
Niche cases are important too!  They were certainly the bane of my debugging efforts on the red-alert system, the sounds.tbl system, and plenty of other situations.

Removing the `Dock_leader` flag after arrival is the perfect solution, and is in fact something I thought was part of the code already.  I've amended the PR with the change.